app/src/main/java/me/phh/sip/SipMessage.kt

```diff
@@ -0,0 +1,119 @@
+package me.phh.sip
+
+import java.io.BufferedReader
+
+
+/* type definitions */
+
+enum class SipMethod {
+	REGISTER,
+	SUBSCRIBE,
+	INVITE,
+	ACK,
+	CANCEL,
+	BYE,
+	OPTIONS,
+	MESSAGE
+}
+
+data class SipStatusCode(val code: Int)
+
+data class SipHeader(val value: String, val parameters: Map<String, String>)
+
+sealed class SipMessage()
+
+data class SipCommonMessage(
+	val firstLine: String,
+	val headers: Map<String, List<SipHeader>>,
+	// TODO: messageBody needs to handle multipart, including nested multipart,
```

Comment:
Have you seen it used yet?

Reply:
the RFC used it for INVITE and I blindly trusted it was used... but now I'm checking it's not for me, although I've only tried calling outside and have no idea what the server would send me when called from outside

Reply:
FWIW here's what rfc5621 has to say about it:

[4.2](https://datatracker.ietf.org/doc/html/rfc5621#section-4.2).  Mandatory Support for 'multipart' Message Bodies
```
   For all MIME-based extensions to work, the recipient needs to be able
   to decode the multipart bodies.  Therefore, SIP UAs MUST support
   parsing 'multipart' MIME bodies, including nested body parts.
   Additionally, UAs MUST support the 'multipart/mixed' and 'multipart/
   alternative' MIME types.  Support for other MIME types such as
   'multipart/related' is OPTIONAL.

      Note that, by default, unknown 'multipart' subtypes are treated as
      'multipart/mixed'.  Also note that SIP extensions can also include
      'multipart' MIME bodies in responses.  That is why both UACs and
      UASs need to support 'multipart' bodies.

   Legacy SIP UAs without support for 'multipart' bodies generate a 415
   (Unsupported Media Type) response when they receive a 'multipart'
   body in a request.  A UAC sending a 'multipart' body can receive such
   an error response when communicating with a legacy SIP UA that
   predates this specification.

      It has been observed in the field that a number of legacy SIP UAs
      without support for 'multipart' bodies simply ignored those bodies
      when they were received.  These UAs did not return any error
      response.  Unsurprisingly, SIP UAs not being able to report this
      type of error have caused serious interoperability problems in the
      past.
```

I think the actual multipart parsing can be left for when we actually need it, but instead of having users of sip messages explicitly check for content-type (99% won't), just directly built it as a map with content-type as key so if they get something of unexpected type it won't work automatically?

e.g. `val body: Map<String, ByteArray>`

Pro:
 - reasonably easy to use
 - makes checking we're parsing what we expect mandatory

Con:
 - won't allow producing nested multipart if required by some extension
 - won't allow sending multiple parts with same content type